### PR TITLE
Liquid context debugging policy

### LINKED ISSUE #838: expose objects available for liquid expressions and policies
It will be great to be able to see values stored in objects available for policy authors and used in liquid extensions. I mean these objects: 
 - line 5-8 https://github.com/3scale/apicast/blob/master/gateway/src/apicast/policy/ngx_variable.lua
 - plus JWT object
 - any other available in future

I think it will be great to have this in echo policy so user add echo policy into policy chain and user can create liquid expressions or write new policy based on values stored in exposed object.

Also I think it will be great to group all available objects to one root object (i.e. window or document object in Javascript in browsers).

### PR DESCRIPTION
**Pull request by davidor**
Closes #838

### CODE REVIEW

File: gateway/src/apicast/policy/liquid_context_debug/context_content.lua [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+local function value_from(object)
+  if type(object) == 'string' or type(object) == 'number' then
+    return object
+  end
+
+  if type(object) ~= 'table' then return nil end
+
+  local res = {}
```
[the remark this answers is not in the document]
Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/context_content.lua]: It looks a bit better :+1:

File: gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  -- There is a lrucache instance with a table that maps service ids (numbers)
+  -- to hosts. So we can have something like { [123456] = "127.0.0.1" }. This
+  -- is a problem when encoding the table into JSON because it's a table with
+  -- many positions but only one != null. When converting this, cjson raises
+  -- an "Excessively sparse arrays" error:
+  -- https://github.com/efelix/lua-cjson/blob/4f27182acabc435fcc220fdb710ddfaf4e648c86/README#L140
+  -- With this call we just convert that number into a string. Not a big deal
+  -- since nobody is going to use those fields from the lrucache instance in
+  -- liquid.
+  cjson.encode_sparse_array(true)
```
[the remark this answers is not in the document]
Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua]: Right, I didn't notice. It wasn't my intention to set this global setting there.

I'll just probably implement some ad-hoc solution in `ContextContent`. If the key is an integer and is greater than X then convert it to string.

File: gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new(config)
+  return self
+end
+
+function _M.content(_, context)
+  local content = context_content.from(context)
```
Comment (mikz) [inline gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua]: We also should add the ngx_variable, right?

Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua]: You're right. And to do that I needed to modify `ngx_variable` so it behaves as a context (linked list): https://github.com/3scale/apicast/pull/849/commits/822345df1bb601fd5bf00c1e799af2f045e92cb0

File: gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua]: Here's how I implemented this: https://github.com/3scale/apicast/pull/849/files#diff-d126ed9fb83c6261c4bb9fefa80fd029R22

File: gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua (hunk 3 above)
Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/liquid_context_debug.lua]: Also, the policy takes that into account now: https://github.com/3scale/apicast/pull/849/commits/98d3d8a946656f12e1d0a14b7bbfa72352fba58c

File: gateway/src/apicast/policy/liquid_context_debug/context_content.lua [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  return res
+end
+
+local function add_content(object, acc)
+  if type(object) ~= 'table' then return nil end
+
+  -- The context is a list where each element has a "current" and a "next".
+  local current = object.current
```
[the remark this answers is not in the document]
Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/context_content.lua]: That's a good idea, but I'd rather leave it for a future PR as it involves adding the functionality in the `LinkedList` module, test it, and check whether we can use it in some of the existing callers.

File: gateway/src/apicast/policy/liquid_context_debug/context_content.lua [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+-- positions with only one occupied. With arrays like these, 'cjson' raises an
+-- "Excessively sparse arrays" error:
+-- https://github.com/efelix/lua-cjson/blob/4f27182acabc435fcc220fdb710ddfaf4e648c86/README#L140
+-- For example, there is a lrucache instance with a table that maps service ids
+-- (numbers) to hosts. So we can have something like { [123456] = "127.0.0.1" }.
+-- This can be limited using a global setting of the 'cjson' library. However,
+-- we're going to implement an add-hoc solution here so we don't affect the
+-- other modules. We're going to convert those large ints to strings as a
+-- workaround.
+local max_integer_key = 10
```
Comment (mikz) [inline gateway/src/apicast/policy/liquid_context_debug/context_content.lua]: I think arrays with 10 elements are common. Maybe better to set this to a something like 1000 ?

Comment (davidor) [reply gateway/src/apicast/policy/liquid_context_debug/context_content.lua]: OK. I'll raise the limit to 1k. I think that in practice large service IDs are going to be the problem.

**Pull request merged**